### PR TITLE
8318709: Improve System.nanoTime performance on Windows

### PR DESCRIPTION
**Pull request by djelinski**
- use LARGE_INTEGER.QuadPart instead of assembling the jlong from high/low parts
- precalculate counts_per_nano to avoid costly floating-point division in counter to nanosecond conversion

Benchmark before:
SystemTime.nanoTime                avgt   15  19,366 � 0,383  ns/op

After:
SystemTime.nanoTime                avgt   15  15,812 � 0,385  ns/op

Tier1-2 clean.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8318709](https://bugs.openjdk.org/browse/JDK-8318709): Improve System.nanoTime performance on Windows (**Enhancement** - P4)

### Reviewers
 * [Conor Cleary](https://openjdk.org/census#ccleary) (@c-cleary - Committer) ⚠️ Review applies to [50bf0dd3](https://git.openjdk.org/jdk/pull/16336/files/50bf0dd369b91c6451d4b328576196300c85da0a)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16336/head:pull/16336` \
`$ git checkout pull/16336`

Update a local copy of the PR: \
`$ git checkout pull/16336` \
`$ git pull https://git.openjdk.org/jdk.git pull/16336/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16336`

View PR using the GUI difftool: \
`$ git pr show -t 16336`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16336.diff">https://git.openjdk.org/jdk/pull/16336.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16336#issuecomment-1776916440)

### CODE REVIEW

**Comment by dholmes-ora** [comment]
Hi @djelinski , I hope to review this tomorrow but need to do some research first. :)

**Review by c-cleary — APPROVED** [review]
LGTM. Can see that `QueryPerformanceFrequency()` sets count the frequency of the performance counter thats fixed at boot too which seems hardware specific, nice gain

**Review by dholmes-ora — COMMENTED** [review]
Looking at:

https://learn.microsoft.com/en-us/windows/win32/api/winnt/ns-winnt-large_integer-r1

It seems the QuadPart needs native 64-bit compiler support, which I'm unclear is available when building for 32-bit Windows (which IIRC is a deprecated port but still present in this release). Also the web page indicates it is only available from Windows 10, which doesn't make a lot of sense.

To be honest I am struggling to understand how LARGE_INTEGER gets used by QPC/QPF as it is a union, so we have to know which member QPC/QPF sets, so we can read back the correct member. Though I wonder if given endian-ness the layout of the structs simply coincides with that of a native 64-bit variable?

**Comment by djelinski** [comment]
Visual Studio supports 64bit integer types even in 32 bit mode. In fact, the JDK requires a compiler with 64bit integer support - it's used in jlong typedef, for example.

I checked the typedef for LARGE_INTEGER in Windows SDK, and it looks exactly like the one on the MSDN page - no ifdefs to check if we are compiling in 64bit mode.

> Though I wonder if given endian-ness the layout of the structs simply coincides with that of a native 64-bit variable?

That's it exactly. 

By the way, here's a nice article from Raymond Chen about this:
https://devblogs.microsoft.com/oldnewthing/20040825-00/?p=38053

**Comment by dholmes-ora** [comment]
Hmm that blog ends with:

> Exercise: Why are the LARGE_INTEGER and ULARGE_INTEGER structures not affected? 

And I'd like to know why we do not have the same alignment issue? (This all seems rather hackish but presumably their library guys get the okay from their compiler guys ... though I wonder then about folk trying to build the Windows code with gcc?

**Comment by djelinski** [comment]
LARGE_INTEGER does not have the same alignment issue because unions are aligned to the alignment required by the largest type, and LONGLONG (aka int64) QuadPart requires 8-byte alignment on 64bit machines.

**Comment by dholmes-ora** [comment]
Ah I see. Okay. This still seems like a code smell on the MS side but that's not relevent to this PR.

File: src/hotspot/os/windows/os_windows.cpp [hunk 1]
```diff
@@ -801,20 +801,12 @@ void os::free_thread(OSThread* osthread) {
...
-
-
-jlong as_long(LARGE_INTEGER x) {
-  jlong result = 0; // initialization to avoid warning
-  set_high(&result, x.HighPart);
-  set_low(&result, x.LowPart);
-  return result;
-}
-
+static double counts_per_nano; // NANOSECS_PER_SEC / performance_frequency
```
Comment (dholmes-ora) [inline src/hotspot/os/windows/os_windows.cpp:804]: Doesn't that calculate nanos_per_count?

Comment (dholmes-ora) [reply src/hotspot/os/windows/os_windows.cpp:804]: Which is what we want of course.

**Review by dholmes-ora — APPROVED** [review]
Looks good. Thanks for this optimization!

**Comment by djelinski** [comment]
Thanks for the reviews!

/integrate

**Pull request closed without merging**